### ISSUE TITLE
False positive: `not-callable` for property that returns a lambda function conditionally

### ISSUE THREAD

**Issue by tushar-deepsource** [PyCQA/pylint#5931]
### Bug description

```python
class C:
    def __init__(self):
        self._x = None

    @property
    def foo(self):
        if self._x is None:
            self._x = lambda: None
        return self._x

c = C()
c.foo()
```

### Command used

```shell
pylint myfile.py | grep E1102
```

### Pylint output

```shell
myfile.py:12:0: E1102: c.foo is not callable (not-callable)
```

### Expected behavior

No error. The code runs when run under python.

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.3
Python 3.10.1 (main, Dec  6 2021, 22:25:40) [Clang 13.0.0 (clang-1300.0.29.3)]
```

### OS / Environment

MacOS Monterey

**Comment by DanielNoord** [comment]
This requires understanding of control flow as we can't infer that the return value of `foo()` is a function easily.

Note that `c._x()` won't get you a false positive, so to avoid the warning in the meantime you might want to explore if that's a possibility. _Assigning `lambda: None` instead of `None` in the `__init__` could work?_

**Comment by DanielNoord** [comment]
Yeah, the code that checks this relies on the `property` decorator and then checks if all return values are callable. Since `None` is not callable (and doesn't get removed due to lacking control-flow) we warn that the call is "not callable".

**Comment by tushar-deepsource** [comment]
If control flow wasn't there, the 2nd example should've also raised E1102 right?

**Comment by tushar-deepsource** [comment]
Found this:

https://github.com/PyCQA/pylint/blob/5324fecb32b0c86a70c88fc196360d721085fdfc/pylint/checkers/typecheck.py#L1231-L1242

The inference seems to work just fine. Here, `attr.infer_call_result` gives us both `Const.None` and `Lambda` objects.

But since we're checking if `all` the inferred results are callable, E1102 is getting raised.

Even if some of the inferred results are callable, pylint reports that it is not callable. Instead of that, maybe it should raise "possibly-not-callable"?

**Comment by tushar-deepsource** [comment]
Proposed solution:

- If we were able to infer call results, and none of the inferred values were callable, raise `not-callable`.
- If we were able to infer call results, and some of the inferred values were callable, raise `possibly-not-callable`. 
- If we were not able to infer call results, do nothing

**Comment by DanielNoord** [comment]
@tushar-deepsource What about changing `possibly-not-callable` to another lint: `assigning-callable-to-non-callable` or something like that?
That seems to be a bigger issue imo?

**Comment by tushar-deepsource** [comment]
@DanielNoord sounds good to me! Did you mean "assigning non callable to callable"?

**Comment by DanielNoord** [comment]
Yeah! I think we should exclude `None` for now but assigning both a string and a callable to the same variable seems problematic and could be a lint for the Refactoring checker I think.

**Comment by tushar-deepsource** [comment]
I'll make a new issue for that.

For this issue, is it okay to not raise `not-callable` in this case then? Mind if I make a PR that only raises `not-callable` only when no inferred values are callable?

**Comment by DanielNoord** [comment]
@tushar-deepsource I think that sounds like the right way forward!
